A major update of the code structure.

Jurwen opened the pull request:
- Updated the Eigen support for matrix manipulation.
- Optimized the code so that definitions and computations are taking less time. 
- Removed the timer for speed up.
- Restructure the code base so it's more logical.
- updated the comments.

Testing on examples of Figure 2, 10, 11, 17 so that their meta data (number of tets, active tets, radius ratio, patches, shells, cells) match up. Will add a unit test support to include these examples and other simple tests.

Jurwen committed:
A major update of the code structure.

- Updated the Eigen support for matrix manipulation.
- Removed the timer for speed up.
- Restructure the code base so it's more logical.
- updated the comments.

qnzhou commented on src/subdivide_multi.h:
```diff
@@ -0,0 +1,302 @@
+
+
+
+//const Eigen::Matrix<double, 16, 4> linear_coeff {{2, 1, 0, 0}, {2, 0, 1, 0}, {2, 0, 0, 1}, {0, 2, 1, 0},{0, 2, 0, 1}, {1, 2, 0, 0}, {0, 0, 2, 1}, {1, 0, 2, 0},{0, 1, 2, 0}, {1, 0, 0, 2}, {0, 1, 0, 2}, {0, 0, 1, 2},{0, 1, 1, 1}, {1, 0, 1, 1}, {1, 1, 0, 1}, {1, 1, 1, 0}};
```
Please remove commented out code. You can alway use git to recover old code if necessary.

qnzhou commented on src/subdivide_multi.h:
```diff
@@ -0,0 +1,302 @@
+///
+bool load_csgTree(const std::string filename, llvm_vecsmall::SmallVector<csg_unit, 20>& tree);
+
+/// an iterative algortihm that traverses through the csg tree
+///
+///@param[in] csgTree           The CSG structure: a list of csg units
+///@param[in] curNode           The current index in the csg structure
+///@param[in] funcInt           The intervals of all the functions
+///
+///@param[out] std::pair
+std::pair<std::array<double, 2>, llvm_vecsmall::SmallVector<int, 20>> iterTree(const llvm_vecsmall::SmallVector<csg_unit, 20>csgTree,const int curNode,const llvm_vecsmall::SmallVector<std::array<double , 2>, 20> funcInt);
```
What are the returned values of this function? The documentation is not very clear.

qnzhou commented on src/subdivide_multi.h:
```diff
@@ -0,0 +1,302 @@
+///
+/// Stores the index of permutations of n less than `funcNum` for the use in `subTet` and `subMI` functions.
+///
+extern llvm_vecsmall::SmallVector<llvm_vecsmall::SmallVector<llvm_vecsmall::SmallVector<std::array<int, 4>, 100>, 3>, 20> multiple_indices;
+
+/// Definations of varialbes before the computations to avoid repetitive defnitions.
+extern int sub_call_two;
+extern int sub_call_three;
+extern int GLOBAL_METHOD;
+extern bool curve_network;
+extern llvm_vecsmall::SmallVector<csg_unit, 20> GLOBAL_CSGTREE;
+extern llvm_vecsmall::SmallVector<std::array<double, 20>, 20> valList;
+extern llvm_vecsmall::SmallVector<std::array<double, 16>, 20> diffList;
+extern llvm_vecsmall::SmallVector<bool, 20> activeTF;
+extern llvm_vecsmall::SmallVector<std::array<double , 2>, 20> funcInt;
+extern llvm_vecsmall::SmallVector<std::array<double, 3>, 20> gradList;
+extern size_t funcNum;
```
Why are these global variables necessary? The subdivision checks should be embarrassingly parallel (ideally run in parallel), there should be no shared data/state between different checks.

Jurwen replied:
This is for the sake of improving the efficiency in the current mode of sequential computing. By defining those variables once, they will not be defined again in each check function millions of times. This is how the current timing is improved. I will restructure everything so that it's ready for the parallel mode.

qnzhou replied:
If the only purpose for these global variables is to avoid repeated memory allocation in sequential computation, they should not be in the global scope. Using local variable with some book-keeping should suffice.  In fact, there should be no global variables in our code (see [here for reasons](https://luzkan.github.io/smells/global-data)).

qnzhou commented on src/subdivide_multi.h:
```diff
@@ -0,0 +1,302 @@
+/// returns a `bool` value that `true` represents positive and `false` represents negative of the input value `x`.
+bool get_sign(double x);
+
+/// returns the dot product of input arrays `a` and `b`.
+double dot(const std::array<double, 3> &a, const std::array<double, 3> &b);
+
+/// returns the dot product of input arrays `a` and `b`.
+double dot(const std::array<double, 2> &a, const std::array<double, 2> &b);
+
+/// returns the dot product of input SmallVector `a` and `b`.
+double dot(const llvm_vecsmall::SmallVector<double, 20> &a, llvm_vecsmall::SmallVector<double, 20> &b);
+
+/// returns the norm of the input array `a` where it represents a vector in 3D.
+double norm(const std::array<double, 3> &a);
+
+/// returns a vector in space by subtracting a 3D coordinate `p2` from `p1`.
+std::array<double, 3> getVec(const std::array<double, 3> &p1, const std::array<double, 3> &p2);
```
These functions should not be necessary when using Eigen.

qnzhou commented on src/subdivide_multi.h:
```diff
@@ -0,0 +1,302 @@
+/// returns a perpendicular vector for the 2D vector `a`.
+std::array<double, 2> perp(const std::array<double, 2> &a);
+
+/// returns the cross product of 3D vectors `a` and `b`.
+llvm_vecsmall::SmallVector<double, 20> cross(const std::array<double, 3> &a, const std::array<double, 3> &b);
+
+/// returns the determinant of the matrix {vec1, vec2, vec3}. Under testing, such computation is faster than turning arrays into `Eigen::Vec3d`
+double det(const std::array<double, 3>& vec1,
+           const std::array<double, 3>& vec2,
+           const std::array<double, 3>& vec3);
+/// returns the determinant of the matrix {vec1, vec2}
+double det(const std::array<double, 2>& vec1,
+           const std::array<double, 2>& vec2);
+
+/// transforms the input of errors at 20 bezier control points for two functions into the correct format that `convex_hull_membership` library can use.
+std::array<double, 40> transpose2d(const std::array<std::array<double, 20>, 2>& matrix);
+
+/// transforms the input of errors at 20 bezier control points for three functions into the correct format that `convex_hull_membership` library can use.
+std::array<double, 60> transpose3d(const std::array<std::array<double, 20>, 3>& matrix);
+
+/// previous matrix multiple function. Now it's replaced by `eigen` package.
+llvm_vecsmall::SmallVector<llvm_vecsmall::SmallVector<double, 20>, 20> matrixMultiply(const llvm_vecsmall::SmallVector<llvm_vecsmall::SmallVector<double, 20>, 20> &matA, const llvm_vecsmall::SmallVector<llvm_vecsmall::SmallVector<double, 20>, 20> &matB);
```
These functions should not be necessary when using Eigen.

qnzhou commented on src/subdivide_multi.h:
```diff
@@ -0,0 +1,302 @@
+#include <Eigen/Sparse>
+
+
+/// Two checks for each grid element: zero-crossing check and distance check. This function performs these two checks under the setting of implicit arrangement(IA), constructive solid geometry(CSG) and their curve networks.
+///
+/// @param[in] pts          4 arrays of three-tuples represent the coordinate of tet vertices. Each subarray represents a coordinate.
+/// @param[in] vals         functions' values at each vertex. Each subarray represents one function's values at four vertices.
+/// @param[in] grads            functions' gradients at each vertices. Each subarray represents one functions' gradients at four vertices and in each axis-aligned partial derivative.
+/// @param[in] threshold            The user-input epsilon value for the distance check.
+/// @param[out] active          A `bool` represents whether the tet is containing part of the geometry, i.e., this tet passes the zero-crossing test.
+///
+/// @return         A `bool` represents whether the tet is "refinable", i.e., passing the zero-crossing test and contains error greater than `threshold`.
+///
+///
+bool subTet(const std::array<std::array<double, 3>,4> &pts,
+            const llvm_vecsmall::SmallVector<std::array<double,4>, 20> &vals,
+            const llvm_vecsmall::SmallVector<std::array<std::array<double, 3>,4>, 20> &grads,
+            const double threshold,
+            bool& active);
```
I would split this function, `subTet`, and the following `subMI` into three smaller functions for better clarity:
1. Build proxy
2. Zero cross check (for dim=2, 1, 0)
3. Distance check for (for dim=2, 1, 0)

Each function only do [one thing](https://cleancoders.com/episode/clean-code-episode-3). With these three smaller building block functions, the logic of IA, CSG, MI and Curve subdivision criteria should be much easier to implement.

qnzhou commented on src/subdivide_multi.h:
```diff
@@ -0,0 +1,302 @@
+};
+
+
+///Defines the enumeration of CSG operations. In the data structure, Intersection is 0, Union is 1, and Negation is 2.
+struct csg_unit{
+    int operation;
+    std::array<int, 2> elements;
+};
+
+///load the csg file
+///
+///@param[in] filename          The name of the input CSG tree
+///@param[out] tree         The loaded tree structure
+///
+bool load_csgTree(const std::string filename, llvm_vecsmall::SmallVector<csg_unit, 20>& tree);
```
CSG logic should be moved to a different file as it is orthogonal to subdivision criteria.

Jurwen commented:
I will change the codebase accordingly. I will have some unit tests ready to verify that the change will not affect the result.

Jurwen committed:
Major code refactor

- modularize the functionality between codes.
- Now, Given an initial grid and a function file, we can produce the output grid with cleaner API.
- CSG has cleaner API.

qnzhou commented on app/gridgen.cpp:
```diff
@@ -0,0 +1,124 @@
+    if (args.grid_file.find(".json") != std::string::npos){
+        grid = grid_mesh::load_tet_mesh(args.grid_file);
+        mtet::save_mesh("init.msh", grid);
+        grid = mtet::load_mesh("init.msh");
```
Why load/save/load??

qnzhou commented on app/gridgen.cpp:
```diff
@@ -0,0 +1,124 @@
+        grid = mtet::load_mesh(args.grid_file);
+    }
+    
+    double max_elements = args.max_elements;
```
`max_elements` should be of type `int`.

qnzhou commented on src/refine_crit.cpp:
```diff
@@ -0,0 +1,645 @@
+/// returns a `bool` value that `true` represents positive and `false` represents negative of the input value `x`.
+bool get_sign(double x) {
+    return (x > 0) ? 1 : 0;
+}
```
This function can be replaced by just `x > 0`.

Jurwen committed:
Update function API and comments

- Modularize the function evaluation into a lambda function. This function can be replaced with other implicit function evaluation.
- Updated comments for all header files, and local functions.

Jurwen commented:
A complete refactor of the codebase: 

- Modularize the grid generation function such that it takes in an initial grid file, a lambda function, and other configuration settings. The output is the final adaptive grid.
- The refinement criteria is factored into three functions that check for three different type of implicit complexes. (`critIA` `critCSG` `critMI`)
- implicit function evaluation and CSG traversing are now lambda functions. They can be replaced with other functions that follow the same API. 
- All the criteria functions use `Eigen` to perform math and matrix operations. 
- The IO functions are grouped together into io files. 
- Created a sheet of unit tests, one for each implicit complex.

qnzhou commented:
Thanks Owen! The code does not compile for me. Here are the errors:

```
/Users/qzhou/Research/Adaptive-grid-generation/app/gridgen.cpp:83:9: error: use of undeclared identifier 'curve_network'
        curve_network = true;
        ^
/Users/qzhou/Research/Adaptive-grid-generation/app/gridgen.cpp:127:27: error: use of undeclared identifier 'curve_network'
    if (!gridRefine(mode, curve_network, threshold, alpha, max_elements, funcNum, implicit_func, csg_func, grid, metric_list, profileTimer))
                          ^
2 errors generated.
```

Can you fix it?

Jurwen committed:
Fixed file and output bug

- changed definitions of local variables
- Added output file options

Jurwen commented:
Sorry, it's fixed now.

qnzhou commented on src/grid_refine.h:
```diff
@@ -0,0 +1,49 @@
+///
+/// @param[in] mode         The modality of the implicit complex, including Implicit Arrangement(IA), Contructive Solid Geometry(CSG), Material Interface(MI).
+/// @param[in] curve_network            Whether we're only computing the curve network of the complex.
+/// @param[in] threshold            The user-defined errro threshold. The smaller the value, the more refined the output grid.
```
Typo: `errro` -> `error`

qnzhou commented on src/grid_refine.h:
```diff
@@ -0,0 +1,49 @@
+/// @param[in] mode         The modality of the implicit complex, including Implicit Arrangement(IA), Contructive Solid Geometry(CSG), Material Interface(MI).
+/// @param[in] curve_network            Whether we're only computing the curve network of the complex.
+/// @param[in] threshold            The user-defined errro threshold. The smaller the value, the more refined the output grid.
+/// @param[in] alpha            An experimental variable. This helps to improve the quality of the tet that contains a part of the iso-surface. In the paper, the default is set to be infinity. If it's set to be 1, it will generate perfect tet in our experiments. However, there is no theoretical gaurantee.
```
I think it is a bit misleading to claim it generates "perfect tet"... You should be more specific on the metric used instead.

qnzhou commented on src/grid_refine.cpp:
```diff
@@ -0,0 +1,251 @@
+    }
+    // write mesh and active tets
+    mtet::save_mesh("tet_grid.msh", grid);
+    mtet::save_mesh("active_tets.msh", grid, std::span<mtet::TetId>(activeTetId));
```
We should not save the tet grid in this function. This function only generates the grid, and it should be up to the user to determine whether to save it.

qnzhou commented on src/grid_refine.h:
```diff
@@ -0,0 +1,49 @@
+/// @param[in] alpha            An experimental variable. This helps to improve the quality of the tet that contains a part of the iso-surface. In the paper, the default is set to be infinity. If it's set to be 1, it will generate perfect tet in our experiments. However, there is no theoretical gaurantee.
+/// @param[in] max_elements         The maximum number of grid elements to cap the algorithm.
+/// @param[in] funcNum          The number of functions that consist the complex.
+/// @param[in] func         The lambda function that computes implicit funcitons' values and gradients given a 3D coordinate and the number of functions.
```
Type: `funciton` -> `function`

qnzhou reviewed (commented):
Thanks Owen! It is getting better. Here is a list of some quick fix requests.

Jurwen committed:
Quick fix of small issues

- Quick fixes of requests from James.

Jurwen committed:
Updated shader example

Jurwen committed:
Updated robustness issue for determining active tet

- Added hash collision detection during the hash map creation. Now, the boolean of whether a tet is active is correct.
- Removed pre-computation of permutations for all functions. Moved to each check.

Jurwen committed:
Update grid_mesh.h

- Using longest edge bisection to initialize grid from Kuhn's subdivision.

Jurwen committed:
Update gridgen.cpp

Deleted unnecessary save/load.

Jurwen commented:
All the requested changes have been corrected.

Jurwen committed:
Fixed initial grid generation

Jurwen committed:
Updated bugs using unnecessary variables

Jurwen committed:
Updated Eigen issues

Changed `parse_convex_points2d` and `parse_convex_points3d` into simpler form.

Jurwen committed:
Aligned scripts with current codebase update

Jurwen committed:
Fixed the bug that caused abort using MI

Jurwen committed:
Update README.md

Jurwen committed:
Update README.md

Jurwen committed:
Update README.md

Jurwen committed:
Update README.md

Jurwen committed:
Experiment robust scripts
